### PR TITLE
command: adds several commands/properties for secondary subs

### LINKED ISSUE #8343: Add properties for the start and end time of secondary subtitles
### Expected behavior of the wanted feature
I have a script which uses `sub-start` to jump to the beginning of the currently playing line. It also factors in the `sub-delay` but it looks like a separate delay for the secondary subs has already been requested in another issue. I would like to have `secondary-sub-start` and `secondary-sub-end` so that I can do the same thing with the secondary subs. For context, I watch videos with Japanese as the primary subs and English as the secondary subs. There are times when only the English subtitles are showing because they are a translation of text that is already visible on screen. Since there is no point in duplicating that text in the Japanese subtitles, `sub-start` is undefined and I can't use my script.

### Alternative behavior of the wanted feature
NA

### Log file
[log.txt](https://github.com/mpv-player/mpv/files/5612074/log.txt)

### PR DESCRIPTION
**Pull request by ripose-jp**
* ~~Adds ```--secondary-sub-delay=<sec>``` as an option and makes it so ```--sub-delay=<sec>``` only affects the primary subtitle.~~
* Adds ```secondary-sub-start``` and ```secondary-sub-end``` properties. Closes #8343.
* Adds a secondary argument to ```sub-seek``` and ```sub-step``` for choosing which subtitle track to seek/step through. The primary subtitle is the default so as not to break compatibility with anything that depended on the old versions.
* Adds documentation for ``secondary-sub-visibility``.

### CODE REVIEW

**Comment by Dudemanguy** [comment]
The code looks good. I don't really use subtitles much so hopefully someone else can weigh in on this.

**Comment by avih** [comment]
What is the use case for secondary subs? Are there any popular players which support more than one sub at a time? I don't think we should add complexity for features which no one uses.

File: player/sub.c [hunk 1]
```diff
@@ -166,9 +166,10 @@ static bool init_subdec(struct MPContext *mpctx, struct track *track)
 
     if (!track->demuxer || !track->stream)
         return false;
-
+
```
Comment (Dudemanguy) [inline player/sub.c:169]: Just noticed this. An extra 4 spaces were added here by accident.

**Comment by avih** [comment]
There are a lot of subtitle options. What makes the options/commands in this PR worthy of supporting, while all the other sub commands/options remain unchanged?

**Comment by ripose-jp** [comment]
I maintain a video player called [Memento](https://github.com/ripose-jp/Memento) specifically geared towards learning Japanese. When the ```secondary-sub-text``` property was added, [I started working on features that depended on it](https://github.com/ripose-jp/Memento/issues/9). Particularly, I put subtitles in a list as they appear so people can read it like a transcript. On first appearance, a subtitle should be added to the list, but if people are seeking around, its important subtitles don't get duplicated. I get around this problem by using the start time as a unique identifier. I can't use the subtitle text itself for this since unique subtitles may contain the same text as a previous one. Since secondary subtitles don't have a ```secondary-sub-start``` property, I've had to approximate it by checking the current position of the video when the ```secondary-sub-text``` property changes and then accounting for subtitle delay. This is non-deterministic, so it leads to issues where some subtitles are ignored as duplicates when they aren't (for example some calling out someone's name in succession), or subtitles getting duplicated if there is enough of a discrepancy between the two approximated start times. Adding the ```secondary-sub-start``` property is really the only easy fix for this problem as far as I can tell.

As for separating subtitle delays, I find the shared delays a nuisance. I've never wanted to retime both the primary and secondary subtitles together, and I can't imagine anyone else really has either.

I don't have any good justification for changing ```sub-seek``` and ```sub-step``` other than ```sub-step``` is useful for retiming subtitles and they share an implementation, so if you change one you need to change both. I'm willing to revert this commit if mpv isn't interested in implementing it, because I don't usually want to retime secondary subtitles anyway.

You're point is valid that it would be better if secondary subtitles had all the same configuration options as the primary subtitle track. I considered the possibly early on when trying to implement these features, but came to the conclusion it was probably more trouble than it's worth. The best way I could think of was giving secondary subtitles their own ```mp_subtitle_opts``` struct, and then refactoring everything around that premise, but since I'm not too familiar with the code base, the risk of messing something up would have been too high.

**Comment by Dudemanguy** [comment]
Adding the properties should be just fine imo. Nothing too controversial there; just exposing some stuff that's already being done internally to the client API. I'm also OK with adding flags to every `sub-*` command to make them work with secondary subtitles. It's basically just extending what's already there. Adding more `--secondary-sub-*` options is where things start to get dicey.

**Comment by ripose-jp** [comment]
The problem with the flag method is that it makes it more difficult to observe properties. I could see it working if subtitle properties were handled the same as [```track-list```](https://mpv.io/manual/master/#command-interface-track-list), but it would obviously break anything that depended on the old way doing things. It's really your call though.

**Comment by avih** [comment]
> The problem with the flag method is that it makes it more difficult to observe properties.

This could be a valid concern. Which such properties did you have in mind which you'd want to observe? (especially considering that `secondary-sub-text` is already separate).

**Comment by ripose-jp** [comment]
At the moment I only need to observe ```secondary-sub-delay``` since I need to update timestamps when it changes.

**Comment by avih** [comment]
Also, if there was a way to "automate" the `secondary-` prefix, while keeping the existing code mostly intact, it could be also be a valid way forward I think.

**Comment by avih** [comment]
> At the moment I only need to observe `secondary-sub-delay` since I need to update timestamps when it changes.

What would change it other than yourself?

**Comment by ripose-jp** [comment]
> What would change it other than yourself?

I'm not sure I follow. The way Memento handles things is that when something changes in mpv, Memento follows. This is why I let people set their own mpv.conf and input.conf's. There's really no way for me to intercept the user changing ```secondary-sub-delay``` without observing it.



> Also, if there was a way to "automate" the `secondary-` prefix, while keeping the existing code mostly intact, it could be also be a valid way forward I think.

This is interesting. I'll look into it.

**Comment by avih** [comment]
By the way, looking at this image from your player https://raw.githubusercontent.com/ripose-jp/Memento/master/example.png , it looks to me like you're rendering the sub yourself (judging by the "selection" at the begining). How does that integrate with mpv to begin with? and where/how does the secondary sub come into play in such case?

I think, if I wanted to implement secondary sub thing, I'd just write a script which parses a `.srt` file and render it whenever/wherever/however I wanted, and send whatever notifications on anything I wanted.

**Comment by avih** [comment]
Also, I think there could be a breaking change in this PR.

I presume (I never used secondary sub nor am I familiar with the relevant mpv code) that all the new options at this PR currently work on both primary and secondary subs, so if we take this PR, these same "old" options would now work only on the primary sub? For instance `sub-delay`.

**Comment by Dudemanguy** [comment]
Only `--secondary-sub-delay` is a breaking change since it causes `--sub-delay` to not affect secondary subtitles anymore. The commands are unaffected as they still default to the primary one (which was previously hardcoded).

**Comment by ripose-jp** [comment]
The easiest way I found to make subtitles hover searchable was to observe the ```sub-text``` property, render the text in a QTextEdit, and then position it over the player. Personally that's fine for me, but [there are people that prefer to have mpv render the subtitles instead](https://github.com/ripose-jp/Memento/issues/12), so I leave the option available.

The use for the secondary subtitle is more apparent in this picture https://user-images.githubusercontent.com/72582120/122622746-91f96800-d04e-11eb-835b-67a8559ba7da.png. The table in the sidebar updates progressively by observing both the ```sub-text``` and ```secondary-sub-text``` properties.

I had considered the script method early on, but since 90% of the time secondary subtitles are baked into mkv or mp4 files it's not a trivial task.

> Also, I think there could be a breaking change in this PR.

It definitely is a breaking change, but I don't see any way around it.
I also can't think of a use case where anything would depend on applying ```sub-delay``` to the secondary subtitle for mpv versions <= 0.33.1 since the only option for interacting with secondary subtitles was ```--secondary-sid=```. It's only been since the last release that [```secondary-subtitle-visibility```](https://github.com/mpv-player/mpv/pull/8462) and ```secondary-sub-text``` have been added.
I would break current versions of Memento, but that's not a big deal since most people can't interact with these features yet anyway, so I have plenty of time to fix it.

**Comment by avih** [comment]
> It definitely is a breaking change, but I don't see any way around it.

Then we need to think about it carefully.

> It's only been since the last release that [`secondary-sub-visibility`](https://github.com/mpv-player/mpv/pull/8462) and ...

Hmmm.. that's not even documented.

> for mpv versions <= 0.33.1 since the only option for interacting with secondary subtitles was `--secondary-sid=`

Exactly. And since then secondary sub things were added in an un-coordinated way, half documented.

And this PR adds more of those (didn't look closely but I assume the docs are good).

I just think that by this stage (and before a next release) all the things related to secondary sub should be unified in some way. I don't know exactly how, and I don't know to what extent, but I think that it's allowed to break recent things (added after 0.33), and much preferably it shouldn't break things which worked at 0.33 .

I just don't like the "random" addition of secondary-sub things in an uncoordinated way.

**Comment by avih** [comment]
For instance, we'd need to list all the subtitles options/properties/commands, identify which of those would be useful also for secondary subs, identify which of those would be reasonably easy to support for secondary subs, and decide how relate to primary/secondary/both subs in a consistent way, including with property observation.

It needs some thought and work, for sure, but I don't think that we should be adding more before we have a good overview of the whole domain and its aspects, even if the next stage would be to implement only a subset of the final goal.

**Comment by ripose-jp** [comment]
> Then we need to think about it carefully.

I propose reworking subtitle properties to work like [`track-list`](https://mpv.io/manual/master/#command-interface-track-list), and using things like ```sub-start``` as aliases for things like ```sub/0/start```. This resolves my observability concerns and makes things more extensible. I don't know how difficult it would be to implement, but that's the best solution I can think of.

It seems unlikely this pull request is going to get merged in its current state. Would you prefer if I closed it?

**Comment by avih** [comment]
> I propose reworking subtitle properties to work like [`track-list`](https://mpv.io/manual/master/#command-interface-track-list), and using things like `sub-start` as aliases for things like `sub/0/start`

That sounds reasonable to me, preferably in a way which doesn't conflict or modify how related options behave in 0.33, i.e. I think it would be fine as an additional thing.

> It seems unlikely this pull request is going to get merged in its current state. Would you prefer if I closed it?

I don't know how likely, so at least for now I'd suggest to keep it open. The above were my personal opinions. Other mpv devs may think differently, and if they think this PR is fine without unifying the secondary sub things, then they can merge it and I won't object.

Unfortunately, so far I don't think I got a clear understanding of what others think about the approach here vs making it unified.

I do hope they speak up their mind.

**Comment by Dudemanguy** [comment]
You could make it work like `track-list` if you want. Subs are just a specific kind of track so I'm not sure how hard/easy that would be. That wouldn't give you `--secondary-sub-delay` though.

**Comment by ripose-jp** [comment]
> Also, if there was a way to "automate" the `secondary-` prefix, while keeping the existing code mostly intact, it could be also be a valid way forward I think.

I've been messing around to see how feasible this is, and I've come to the conclusion that the secondary subtitles can inherit almost all of the code used for primary subtitles just by giving the secondary subtitles their own ```m_sub_options```. I've been able to accomplish this pretty easily by just changing
https://github.com/mpv-player/mpv/blob/b3fccf080369730e1b4d6a3435c90124a74566f7/options/options.h#L186
into ```struct mp_subtitle_opts *subs_rend[2];``` and then passing in the index into that array via the ```priv``` variable. Here's an example of the macro I've been using:
image

Getting options to work hasn't been as pretty since I basically just duplicated this [struct](https://github.com/mpv-player/mpv/blob/b3fccf080369730e1b4d6a3435c90124a74566f7/options/options.c#L235) and prepended ```secondary-``` to everything there. I don't think there's any macro I could write that would automate that process unfortunately.

Besides a few tweaks to a few methods here and there, that's all I've had to do though.

While the breakage concerns stemming from decoupling the primary and secondary subtitles still remains, I still think method is the way forward, and not this PR. I'll probably leave this open until I'm certain I can make the code I'm working on good enough for a pull request.

**Comment by avih** [comment]
> nto `struct mp_subtitle_opts *subs_rend[2];` and then passing in the index into that array via the `priv` variable. Here's an example of the macro I've been using

It's not obvious to me where this macro is used. Can you post the complete patch someplace for applying your approach to the first three options `sub-delay`, `sub-fps` and `sub-speed` ?

> I basically just duplicated this [struct](https://github.com/mpv-player/mpv/blob/b3fccf080369730e1b4d6a3435c90124a74566f7/options/options.c#L235) and prepended `secondary-` to everything there

Not sure how overlapping this suggestion is, but maybe can we could replace for instance this:
```c
{"sub-delay", OPT_FLOAT(sub_delay)},
```

with this?
```c
{SUBT_OPT("sub-delay", OPT_FLOAT(sub_delay))},
```

Where `SUBT_OPT` maps 3 option names to 3 functions: `sub-delay`, `primary-sub-delay` and `secondary-sub-delay`.

The three functions implementations would be automated too in some way, where the non-prefixed name applies the value to both, while the prefixed names apply only to one of those?

We'd still need to think how reading the properties work, as obviously setting `sub-delay' for instance could be made to affect both subs, but reading/observing is likely to retrieve one.

The main issue is probably with the non-prefixed name, where I think that on read it should fetch only the primary value?

We'd also need to ensure that the properties->options bridge works.

And a bit beyond the names technicality, did you try to figure out if all the options at `mp_subtitle_sub_opts` and `mp_sub_filter_opts` are applicable to both subs?

What about sub-related properties or options beyond these two structs? e.g. `sub-demuxer`, `sub-auto` and few more, how do they fit, if at all?

And what about commands? can we start by listing all the sub-related commands and evaluate how applicable is it to use such functionality-triplet there too (maybe as an optional final argument argument)?

I.e. evaluate the scope, applicability, implementation approach/impact, and the "template" for user-facing names and behaviour of adding full support for secondary subs.

Then, if the full support requires a lot of changes which we prefer to not do right now or at one go, evaluate if it's still useful and applicable to do that only for some of the options/properties/commands in a way which allows to incrementally make it more complete in the future - but with the "template" for names/behavior/functionality already in place.

It is a lot of work, no doubt, but even if we can't do all of it, we should at least get a feel for what the full scope means in practice, and then we can make better decisions IMO.

**Comment by ripose-jp** [comment]
> It's not obvious to me where this macro is used. Can you post the complete patch someplace for applying your approach to the first three options `sub-delay`, `sub-fps` and `sub-speed` ?

Here's a [patch](https://github.com/mpv-player/mpv/files/6779599/patch.txt). ```secondary-sub-delay``` seems to work fine, but ```secondary-sub-speed``` seems to stops showing subtitles for values outside of 0.81 - 1.19. Not sure how well ```secondary-sub-fps``` works, but its probably similar to ```secondary-sub-speed```. This is really just a proof of concept for how this could work, nothing serious.

> Where `SUBT_OPT` maps 3 option names to 3 functions: `sub-delay`, `primary-sub-delay` and `secondary-sub-delay`.

I can see this being possible by changing all the shared settings in the ```mp_subtitle_opts``` struct into an array of 3, the first value shared, and the second and third values specific to the primary and secondary subtitles respectively. It would required more far reaching changes to the code than the proof of concept, but its probably worth it to avoid duplicating things unnecessarily. It would also make it easier for the primary and secondary subtitles to share settings if there are settings which should be applied to both subtitles in all cases.

Something that may not be able to be automated/needs more care is adding values to ```property_osd_display``` in command.c since capitalization is a concern.

> The main issue is probably with the non-prefixed name, where I think that on read it should fetch only the primary value?

I think that a non-prefixed name should only fetch the shared value. ```primary-``` and ```secondary-``` should retrieve subtitle specific values. So in order to compute the effective delay, something like ```sub-delay + primary-sub-delay``` needs to be done in code. This might sounds tedious, but this is already how it needs to be done in 0.33.1 since ```sub-start``` fetches the start time without delay, so there is already precedent.

> And a bit beyond the names technicality, did you try to figure out if all the options at `mp_subtitle_sub_opts` and `mp_sub_filter_opts` are applicable to both subs?

I haven't messed with `mp_sub_filter_opts` at all, so I couldn't say, but almost all of the `mp_subtitle_sub_opts` opts work. Some of them have weird behaviors like `sub-pos` which appear to work for bitmap secondary subtitles, but not for text-based secondary subtitles. [That behavior alone is enough to keep it in though since bitmap secondary subtitles have been broken for a long time.](https://mpv.io/manual/master/#options-secondary-sid)

> What about sub-related properties or options beyond these two structs? e.g. `sub-demuxer`, `sub-auto` and few more, how do they fit, if at all?

Just by glancing at them, they probably need to be implemented manually on a case by case basis if they're relevant.

> And what about commands? can we start by listing all the sub-related commands and evaluate how applicable is it to use such functionality-triplet there too (maybe as an optional final argument argument)?

The only commands I can find that are relevant to both the primary and secondary subtitles  separately are ```sub-seek``` and ```sub-step```. Maybe ```screenshot``` if there some reason someone would want one subtitle track to appear but not the other?

**Comment by avih** [comment]
Sorry for the delay.

> Here's a [patch](https://github.com/mpv-player/mpv/files/6779599/patch.txt).

Yeah. Thanks for the work. It doesn't look pretty (i.e. simple), but I guess this kind of scope is unavoidable if we're going to make secondary-sid an equal member rather than an afterthought.

Also, looking at the git history, secondary-sid was added at 2016 and was not touched since (until recently). It was half baked then, and it still is now. The recent additions were not well-integrated, and there's no "system" in place to handle secondary values (in terms of code integration and behavior consistency).

> `secondary-sub-delay` seems to work fine, but `secondary-sub-speed` seems to stops showing subtitles for values outside of 0.81 - 1.19. Not sure how well `secondary-sub-fps` works, but its probably similar to `secondary-sub-speed`. This is really just a proof of concept for how this could work, nothing serious.

Right, for it to actually work well some mechanism would have to be put in place to use all the secondary values when referring to the secondary sid, and considering how many such values there are, I imagine the approach in this patch would be a nightmere to follow.

> Proof of concept [patch](https://github.com/mpv-player/mpv/files/6780021/patch.txt)

E.g. scaling this kind of change to all the dozens of sub-related options is simply not viable - it doesn't scale at all:
```diff
-        if (track->type == STREAM_SUB)
-            return -opts->subs_rend->sub_delay;
+        if (track->type == STREAM_SUB) {
+            float delay = -opts->subs_rend->sub_delay[0];
+            if (mpctx->current_track[1][STREAM_SUB] == track) {
+                delay -= opts->subs_rend->sub_delay[2];
+            } else {
+                delay -= opts->subs_rend->sub_delay[1];
+            }
+            return delay;
+        }
```

> I think that a non-prefixed name should only fetch the shared value.

I probably didn't explain myself well, but what I had in mind (without thinking it through) was not your approach - change struct members to arrays of 3 for primary/secondary/shared, but rather keep the struct unmodified and just add another instance (variable) of it for secondary subs - and without a 3rd instance of shared values.

So that setting primary/secondary sets the value at the respective instance, and unprefixed would set it in both, while reading unprefixed would read from the primary - like it was before (recall, one of the goals is to not break existing behavior).

Also, maybe with two instances of an unmodified struct it would be possible to keep all the current code which sets/uses these values mostly unmodified, and just find a way to make it work with the primary/secondary struct as needed.

For instance, maybe just call the rendering function once for the primary values, and if secondary is used - call it again with the secondary values struct, or some such (again, I didn't try to look how many functions there are which refer to these options).

> The only commands I can find that are relevant to both the primary and secondary subtitles separately are `sub-seek` and `sub-step`

At least this seems to be limited in scope.

Overall, judging by the scope your patch appears to require, it's not scalable and not viable IMO. I don't know if it would be much more viable with my suggested approach, but I can't ask you to spend time trying it out.

So I don't know what to do about this PR. I really don't want to add sporadic secondary options without a consistent system in place - there are dozens of sub options, and where does it stop? On the other hand, making secondary subs fully integrated as equals seems way too much work for the value it gives, and is also a potential for many new bugs.

Ideas?

**Comment by Dudemanguy** [comment]
After some brief IRC discussion on this PR, I'd be okay with merging the two commits that add commands/properties but not the `--secondary-sub-delay` one. The former are pretty self-contained and simple. The sub delay one touches a ton of stuff at once and as discussed earlier raises a lot of questions about the mpv subtitle code. If it ever gets refactored/rewritten, I wouldn't expect the changes to properties/commands to be terribly complicated.

**Comment by ripose-jp** [comment]
> After some brief IRC discussion on this PR, I'd be okay with merging the two commits that add commands/properties but not the `--secondary-sub-delay` one.

This is fine by me. I've updated the PR and commits to do only this.

I've also added another commit documenting ``secondary-sub-visibility`` just to get it out of the way.

**Comment by Dudemanguy** [comment]
>I've also added another commit documenting secondary-sub-visibility just to get it out of the way.

Thanks for doing that one too!

**Comment by nev3rfail** [comment]
> What is the use case for secondary subs? Are there any popular players which support more than one sub at a time? I don't think we should add complexity for features which no one uses.

I am extremely sorry about the necroposting, but just wanted to clarify.
We are here, the users of `secondary-sid` option. I personally know at least three more people that use it. so it is not for no one :)

Secondary subs are great in case audio track is not viewer's native language and viewer is not profound enough to read & understand the native subtitles completely. So the viewer hooks up secondary subs just to instantly see how some word or language construct translates. Also, there are multiple services on the web that allows this kind of functionality, even some payed ones. Like, there are literally services that charge user for allowing secondary subs, and users that are paying for it :)

I found this post because I purposely googled and read mpv's manual on a case to add delay to secondary subs to watch a movie with my friend who is not very good at English. It's a nuisance that `secondary-sub-delay` did not make it into the final merge and that secondary-sub* options are indeed half baked. But even in the current state it works and is used by people. I just wish it to be more feature rich and consistent.

**Comment by andrewrk** [comment]
I tried out the secondary subtitle feature last night for putting English up top to help when I cannot understand the Japanese below. It's great! Sorry for necroposting but I just want to thank everyone for working together on this. I'm a very happy user over here. 皆さん、本当にありがとうございました！

**Pull request merged**